src/app/src/api/referenceImages.js

```diff
@@ -0,0 +1,80 @@
+import { useDispatch } from 'react-redux';
+import { useTrailingDebounceCallback } from '../hooks';
+import api from './api';
+
+const referenceImagesApi = api.injectEndpoints({
+    endpoints: build => ({
+        uploadReferenceImage: build.mutation({
+            query: ({ boundaryId, ...details }) => ({
+                url: `/boundaries/${boundaryId}/reference-images/`,
+                method: 'POST',
+                data: details,
+            }),
+            onQueryStarted: async (
+                { boundaryId },
+                { dispatch, queryFulfilled }
+            ) => {
+                try {
+                    const { data: newImage } = await queryFulfilled;
+
+                    dispatch(
+                        api.util.updateQueryData(
+                            'getBoundaryDetails',
+                            boundaryId,
+                            draftDetails => {
+                                draftDetails.reference_images.push(newImage);
+                            }
+                        )
+                    );
+                } catch {}
+            },
+        }),
+        updateReferenceImage: build.mutation({
+            query: ({ boundaryId, id, ...details }) => ({
+                url: `/boundaries/${boundaryId}/reference-images/${id}/`,
+                method: 'PUT',
+                data: details,
+            }),
+            // no onQueryStarted to update getBoundaryDetails here because
+            // it is handled by the debounced callback in Sidebar.js
+        }),
+    }),
+});
+
+export function useDebouncedUpdateReferenceImageMutation(boundaryId) {
+    const dispatch = useDispatch();
+    const [updateReferenceImage, meta] = useUpdateReferenceImageMutation();
+
+    const debouncedUpdate = useTrailingDebounceCallback({
+        callback: updatedImage => {
+            updateReferenceImage({ boundaryId, ...updatedImage });
+        },
+        immediateCallback: updatedImage => {
+            dispatch(
+                api.util.updateQueryData(
+                    'getBoundaryDetails',
+                    boundaryId,
+                    draftDetails => {
+                        const imageIsImageToBeUpdated = image =>
+                            image.id === updatedImage.id;
+
+                        draftDetails.reference_images =
+                            draftDetails.reference_images.map(image =>
+                                imageIsImageToBeUpdated(image)
```

Comment:
Prefer a simpler name, like `shouldUpdate`

```suggestion
                                shouldUpdate(image)
```